staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -140,6 +155,7 @@ func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getA
 		}
 		buffer.endIndex++
 	}
+	sort.Sort(sortableWatchCacheEvents(buffer.buffer))
```

Comment:
I don't understand why we would want to do that. 
This seems like unnecessary work to me.

I understand that etcd does it, but I believe it's because of how the internal data store is organised. 
So, for etcd, it probably doesn't require any additional effort . . . 

Why do clients care about the order of init events ?

Reply:
We want to provide consistency across implementations - if someone is using etcd implementation and then is switching to watchcache then it may break their flows.

FWIW - https://github.com/kubernetes/kubernetes/pull/113730 is a bug-fix from almost the same category.

Reply:
> if someone is using etcd implementatio

mhm, most of the workloads use the watchcache implementation...

>  ... and then is switching to watchcache then it may break their flows.

mhm, only when this is documented, is this documented ?

also, why not to make a distinction between `watch` and `list` requests ?

Reply:
> mhm, most of the workloads use the watchcache implementation...

Most doesn't matter - if anyone is using the other, we need consistency.

> mhm, only when this is documented, is this documented ?

Not only when it's documented. People are making assumptions based on how the system is working.
The fact that something wasn't documented doesn't mean that our users/customers are less angry when we change the behavior.

> also, why not to make a distinction between watch and list requests ?

etcd implementation for watch-list is returning sorted items - we want the watchcache implementation to do exactly the same.

Also - this is the only way of opening the doors for pagination if it appears to be desired.

Reply:
watch-list as it is implemented today requires clients to observe all items for consistency. So a client might get initial items, followed by any watch items (assuming watchcache was behind the storage). So the order of stream events cannot be guaranteed.  

In general for `list` requests, why not change the implementation so that it stores items in the expected order instead of sorting them on every request? I know that this is a bigger ask but there are so many sorting algs with different tradeoffs for cpu and memory.

Reply:
> watch-list as it is implemented today requires clients to observe all items for consistency. So a client might get initial items, followed by any watch items (assuming watchcache was behind the storage). So the order of stream events cannot be guaranteed.

I don't understand this comment:
(a) this PR is guaranteeing the alphabetical order for the "list" part of watch-list
(b) the further part has a guarantee of delivery by increased RV
So we will have order guarantees.


> In general for list requests, why not change the implementation so that it stores items in the expected order instead of sorting them on every request? 

The cost of sorting the table is negligible comparing to serializing the full objects in that table.
Sure - we can consider optimizing it in the future, but the ROI from this effort would be neligigble.

Reply:
> The cost of sorting the table is negligible

What about memory allocations ? Is this an in-place sort? (The documentation doesn't mention memory cost for `sort.Sort`)

What I meant in my previous comment is that for watchlist request we cannot guarantee the ordering. The following order is possible:
1. the watchcache is behind etcd, say etcd is at rv=5 and the watchcache is at rv=4, so `bookmarkAfterRV=5`
2. we have the `watchCacheEvent` with the following keys (init events): `foo`, `bar`
3. we sort the init events and we get `bar`, `foo`
4. we send the init events to a client
5. at some point we get a new event with the following key `baz` at `rv=5`
6. we send the newly received event to the client followed by the special bookmark event

at the client-side we received events in the following order: `bar`, `foo`, `baz` whereas the desired order was `bar`, `baz`, `foo`.

Reply:
> What about memory allocations ? Is this an in-place sort?

Yeah - i was looking into it in the previous PR I linked and I didn't see any impact there.

> at the client-side we received events in the following order: bar, foo, baz whereas the desired order was bar, baz, foo.

That was the initial proposal of doing that, but it's not what we did. What we did is that watchcache is effectively waiting to be synced to a given RV before even starting sending events:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1302

So we're fine here.

Reply:
> Yeah - i was looking into it in the previous PR I linked and I didn't see any impact there.

would it make sense to benchmark memory usage ? (just to be sure)

> That was the initial proposal of doing that, but it's not what we did. What we did is that watchcache is effectively waiting to be synced to a given RV before even starting sending events:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1302
> So we're fine here.

I think this is true for `requestedWatchRV>0` but not all requests set `requestedWatchRV` to such a value i.e. initial requests from informers (`requestedWatchRV=unset`). 

I will double check that ^

Reply:
Friendly ping

Reply:
Okay, so it looks like we won't wait for the cache to catch up when `requestedWatchRV=0|unset`. Here is the function that does the waiting https://github.com/kubernetes/kubernetes/blob/39bb8f4bb1d013937aceac6c387563ffe13545c5/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L468

And [this](https://github.com/kubernetes/kubernetes/blob/7e35823690df01bd019a88d3346bd3ac820afaca/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L519) is the place where we read the value of `requestedWatchRV` - (it is read only, we don't change that value after reading).

Reply:
I've extended the `RunWatchSemanticInitialEventsExtended ` test to demonstrate the issue (https://github.com/kubernetes/kubernetes/pull/122676)

Reply:
So I think that what we're doing now is bad.

For cases where update/delete events can be send, "order" doesn't make any sense.
This PR fixes the problem that the actual "initial events" are send in alphabetical order. But after that, by definition of watch we have to sort all the remaining events by RV.

If we want the whole stream to be ordered, we have to wait for watchcache to be synced to requested RV first.

Reply:
@p0lyn0mial - with your PR already merged, this should just work now.